src/dir_contents.ml

```diff
@@ -449,44 +449,46 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
 
 module Dir_status = struct
   type t =
-    | Empty_standalone of File_tree.Dir.t option
-    (* Directory with no libraries or executables that is not part of
-       a multi-directory group *)
-
-    | Is_component_of_a_group_but_not_the_root of
-        Super_context.Dir_with_jbuild.t option
-    (* Sub-directory of a directory with [(include_subdirs x)] where
-       [x] is not [no] *)
-
-    | Standalone of File_tree.Dir.t
-                    * Super_context.Dir_with_jbuild.t
-    (* Directory with at least one library or executable *)
+    | Standalone of
+        (File_tree.Dir.t * Super_context.Dir_with_jbuild.t option) option
+    (* Directory not part of a multi-directory group. The argument is
+       [None] for directory that are not from the source tree, such as
+       generated ones. *)
 
     | Group_root of File_tree.Dir.t
                     * Super_context.Dir_with_jbuild.t
     (* Directory with [(include_subdirs x)] where [x] is not [no] *)
 
+    | Is_component_of_a_group_but_not_the_root of
+        Super_context.Dir_with_jbuild.t option
+    (* Sub-directory of a [Group_root _] *)
+
   let is_standalone = function
-    | Standalone _ | Empty_standalone _ -> true
+    | Standalone _ -> true
     | _ -> false
 
   let cache = Hashtbl.create 32
 
-  let analyze_stanzas stanzas =
-    let is_group_root, has_modules_consumers =
-      List.fold_left stanzas ~init:(None, false) ~f:(fun acc stanza ->
-        let is_group_root, has_modules_consumers = acc in
-        match stanza with
-        | Include_subdirs (loc, x) ->
-          if Option.is_some is_group_root then
-            Loc.fail loc "The 'include_subdirs' stanza cannot appear \
-                          more than once";
-          (Some x, has_modules_consumers)
-        | Library _ | Executables _ | Tests _ ->
-          (is_group_root, true)
-        | _ -> acc)
-    in
-    (Option.value is_group_root ~default:No, has_modules_consumers)
+  let get_include_subdirs stanzas =
+    List.fold_left stanzas ~init:None ~f:(fun acc stanza ->
+      match stanza with
+      | Include_subdirs (loc, x) ->
+        if Option.is_some acc then
+          Loc.fail loc "The 'include_subdirs' stanza cannot appear \
+                        more than once";
+        Some x
+      | _ -> acc)
```

Comment:
Maybe it's a bit too early, but I wonder if it will pay off to enforce invariants like this in the type of `Dir_with_jbuild`:

```
type dir_with_jbuild =
  { src_dir : ..
  ; ctx_dir : ..
  ; scope : ..
  ; scope : ..
  : kind : ..
  : executables : Executables.t list
  ; libraries : Library.t list
  ; include_dirs : Include_subdirs.t option
  ..
  etc.
  }
```

It seems like we're iterating over this list of stanzas quite a few times in the code base.